### PR TITLE
udev-rule: rework logging about udev rules

### PR DESCRIPTION
**Pull request by yuwata**

### CODE REVIEW

**Review by ldv-alt — APPROVED** [review]
This PR essentially replaces `log_rule_*` and `log_token_*` with `log_event_*` and `log_line_*`.
The former used to take `UdevRules*` argument, the latter take `UdevRuleToken*` and `UdevRuleLine*` arguments, respectively, which is more consistent.
Also, `log_line_*` have been made more consistent by taking `UdevRuleLine*` argument instead of `UdevRuleFile*` and `line_nr`.

Overall, logging interface looks more consistent now.

File: src/udev/udev-rules.c [hunk 1]
```diff
@@ -365,24 +383,25 @@ UdevRules *udev_rules_free(UdevRules *rules) {
         return mfree(rules);
 }
 
-static int rule_resolve_user(UdevRules *rules, const char *name, uid_t *ret) {
+static int rule_resolve_user(UdevRuleLine *rule_line, const char *name, uid_t *ret) {
+        Hashmap **known_users = &LINE_GET_RULES(rule_line)->known_users;
```
Comment (dtardon) [inline src/udev/udev-rules.c:399]: Missing `ASSERT_PTR`. The var. is dereferenced unconditionally later.

File: src/udev/udev-rules.c [hunk 2]
```diff
@@ -400,24 +419,25 @@ static int rule_resolve_user(UdevRules *rules, const char *name, uid_t *ret) {
         return 0;
 }
 
-static int rule_resolve_group(UdevRules *rules, const char *name, gid_t *ret) {
+static int rule_resolve_group(UdevRuleLine *rule_line, const char *name, gid_t *ret) {
+        Hashmap **known_groups = &LINE_GET_RULES(rule_line)->known_groups;
```
Comment (dtardon) [inline src/udev/udev-rules.c:435]: Missing `ASSERT_PTR` here too.

File: src/udev/udev-rules.c (hunk 1 above)
Comment (ldv-alt) [reply src/udev/udev-rules.c:399]: Is it? `*known_users` is passed down to `hashmap_get` which is expected to handle NULL properly, isn't it?

Comment (yuwata) [reply src/udev/udev-rules.c:399]: In this stage, `known_users` may be NULL, and will be allocated in this function.

File: src/udev/udev-rules.c (hunk 2 above)
Comment (yuwata) [reply src/udev/udev-rules.c:435]: Also, that may be NULL.

**Comment by yuwata** [comment]
Updated. The basic direction is not changed, but many minor adjustments are applied. So, PTAL.

File: src/udev/udev-rules.c [hunk 3]
```diff
@@ -183,175 +185,203 @@ struct UdevRules {
...
                 int _lv = (level);                                      \
-                UdevRuleFile *_f = _r ? _r->current_file : NULL;        \
-                UdevRuleLine *_l = _f ? _f->current_line : NULL;        \
+                sd_device *_dev = (device);                             \
+                UdevRuleFile *_f = (file);                              \
                 const char *_n = _f ? _f->filename : NULL;              \
                                                                         \
-                rule_file_mark_issue(_f, _lv);                          \
+                if (_dev && _f)                                         \
+                        _f->issues |= (1U << _lv);                      \
```
Comment (ldv-alt) [inline src/udev/udev-rules.c]: The `_dev` check introduced in the latest edition doesn't look correct to me: this disables setting of `issues` in all logging based on `log_file_full_errno_zerook`, I suppose it breaks udevadm verify.

Comment (yuwata) [reply src/udev/udev-rules.c]: `udevadm verify` does not process any rules, and `_dev` is always NULL on parsing rules.

Maybe, I should split the change into another commit with description.

Comment (ldv-alt) [reply src/udev/udev-rules.c]: It _does_ break udevadm verify, testsuite-17.11.sh no longer passes.

Comment (yuwata) [reply src/udev/udev-rules.c]: Ah, I understand my failure, the condition must be inverted. Ouch!

Comment (ldv-alt) [reply src/udev/udev-rules.c]: In particular, all `log_file_*` primitives are affected.

**Comment by yuwata** [comment]
Thank you for the comments. Fixed the issue found by @ldv-alt, split the large commit into small pieces, added several cleanups for 'verify' command.

File: src/udev/udevadm-verify.c [hunk 4]
```diff
@@ -125,6 +128,7 @@ static int verify_rules(UdevRules *rules, char **files) {
                         rv = r;
         }
 
+        log_notice("%zu udev rules files are loaded and verified.", strv_length(files));
```
Comment (ldv-alt) [inline src/udev/udevadm-verify.c]: With this change, testsuite-17.11.sh needs to be updated.

Comment (yuwata) [reply src/udev/udevadm-verify.c]: Ah, indeed. I will drop it, at least tentatively, as it is cosmetic.

**Pull request merged**